[WIP][Prism] Structures preferences

warlof opened the pull request:
I would have a feedback on the way I'm currently handling the new feature for Prism and its BP calculator.

I'm attempting to have as much possible a cleaner form (`frmStructurePreference.vb`) in order to avoid useless methods and event handlers.
Todo so, I'm using databinding which are binding form fields to model properties (`StructurePreference.vb`)

The `StructurePreference.vb` model is storing a list of bonus model (`StructureBonus.vb`) with bonuses which are applied by the structure.

Then, when you're simulating an industry job from BP Calculator which is either a manufacturing or a science job, we just need to iterate over bonuses and call `IsAppliedTo()` method.
This method will return true if the bonus from which it is called is applying to the passed item.

Eshva commented on EveHQ.Prism/Classes/StructureBonus.vb:
```diff
@@ -0,0 +1,229 @@
+    ''' <summary>
+    ''' The Time Efficiency Rate provided Bonus
+    ''' </summary>
+    Public Property TimeEfficiency As Decimal
```
It's would be better to separate 2 TE bonuses with 2 properties: ScientificTimeEfficiencyBonus and ManufacturingTimeEfficiencyBonus. Other 2 bonuses: ScientificCostEfficiencyBonus and ManufacturingMaterialEfficiencyBonus. It would be cleaner than the current design with a boolean property IsScientific.

Eshva replied:
Or better, separate 2 bonus kinds with 2 classes: 1 for scientific jobs and 1 for manufacturing.

warlof replied:
I can drop the boolean IsScientific and create two fully distinct property in order to store bonuses ?

I don't really figure what will be in `StructureScienceBonus` and in `ManufacturingBonus` since all methods/properties are shared (except the CE atm, but it could be added by a rigs for manufacturing jobs in a future, so...)

Eshva replied:
1. Don't develop for future.
2. I wrote below they are different bonuses. Don't generalize to much.

Eshva commented on EveHQ.Prism/Classes/StructureBonus.vb:
```diff
@@ -0,0 +1,229 @@
+    ''' <summary>
+    ''' The Group to which the Bonus should be Applied
+    ''' </summary>
+    Public Property Group As String
```
What is the Group? Group of what?

warlof replied:
Replaced by `Name` since it's the name of the bonus

Eshva commented on EveHQ.Prism/Classes/StructureBonus.vb:
```diff
@@ -0,0 +1,229 @@
+    ''' Key indicate item market group ID
+    ''' Value indicate tech level, 0 mean no restriction
+    ''' </summary>
+    Public ReadOnly Property AppliesToMarketGroups() As Dictionary(Of Integer, Integer)
```
Why not to add another class with properties MarketGroupId and ***TechLevel? *** because from your comment I don't understand which kind of restrictions it is. May be ProductMaxTechLevel?

warlof replied:
What about this ?

```
Public Class StructureBonusFilter
    Public MarketGroupId As Integer
    Public TechLevel As Integer ' 0 mean no level restriction, 1 mean all T1, 2 mean all T2
End Class
```

Eshva commented on EveHQ.Prism/Classes/StructureBonus.vb:
```diff
@@ -0,0 +1,229 @@
+                Where group.Key = type.MarketGroupId
+                Where StaticData.GetAttributeDataForItem(type.Id).  ' Fetch item attributes and check tech level (422)
+                    Where(Function(a) a.Key.Equals(422)).
+                    First().
```
It can rise an Exception if there is no such itemType. May be in this place you should detect it and return false?

warlof replied:
From `GetAttributeDateForItem()` method usage you mean ?

Eshva replied:
Yes.

Eshva commented on EveHQ.Prism/Classes/StructureBonus.vb:
```diff
@@ -0,0 +1,229 @@
+    ''' <summary>
+    ''' Seeds bonus array with T1/Faction Frigates, Destroyers and Shuttles
+    ''' </summary>
+    Public Sub ApplyToT1SmallShips()
```
If you will return ```this``` from those methods you can nest execution of those methods: ```bonus.ApplyToT1SmallShips().ApplyToT2SmallShips().ApplyToT3SmallShips()```...

warlof replied:
Also, what do you think about replacing the tech method by generic ones including an optional parameter ?

```
Public Function ApplyToMediumShips() As StructureBonus
    Return Me.ApplyToMediumShips(0)
End Function

Public Function ApplyToMediumShips(techLevel As Integer) As StructureBonus
```

Eshva replied:
Bonuses are different, has different properties and apply to different activities. It's a bad idea to generalize them. Separate classes are better. May be with a common ancestor.
But I think you need to start from another end, from usage of those data in calculations. This way you will not guess and overengineer. You will know what you need.

Eshva commented on EveHQ.Prism/Classes/StructurePreference.vb:
```diff
@@ -0,0 +1,113 @@
+    ''' <summary>
+    ''' The Type of the Structure
+    ''' </summary>
+    Public Property Type() As EveType
```
StructureItemType?

warlof replied:
Well, that's already structure, so `StructureItemType` doesn't make sense.
`ItemType` ?

Eshva commented on EveHQ.Prism/Classes/StructurePreference.vb:
```diff
@@ -0,0 +1,113 @@
+''' <summary>
+''' A Structure Definition
+''' </summary>
+Public Class StructurePreference
```
FavoriteStructure?
Structure can't have preferences.

warlof replied:
Why not `Structure` only ?!

Eshva replied:
May be just Structure but it's a bit misleading because it's so common term. The is up to you.

Eshva commented on EveHQ.Prism/Classes/StructurePreference.vb:
```diff
@@ -0,0 +1,113 @@
+    ''' A list of possible EveType for Structure Type
+    ''' </summary>
+    ''' <returns></returns>
+    Public Function GetPossibleTypes() As List(Of EveType)
```
It shouldn't be here. You need this list only in UI.

Eshva commented on EveHQ.Prism/Forms/frmBPCalculator.vb:
```diff
@@ -1765,6 +1765,15 @@ Namespace Forms
+        '
+        Private Sub btnEditStructPreference_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnEditStructPreference.Click
+            Dim formStrucPref As frmStructurePreference = New frmStructurePreference()
+            formStrucPref.Show()
```
Please don't use abbreviations in names.

warlof committed:
WIP feature for structure bonus on Prism BP Calculator
add missing designer for bpcalculator
add more category seed and few code refactorying
code refactor and add components, structures, starbases & fuel blocs seeders
move UI DataSource loaders to form
fix typo
make filter application method returning object itself in order to allow thoses methods to be chained
start UI bonus implementation
one more
switch ME/CE/TE rate to a dictionary using system security as key
move class to dedicated namespace
rename structure to upwellstructure
Add ME/TE/CE multiplicators into bonus class and update properties.
Add a new method which will return the computed bonus rate according to the system.
